doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,84 @@
+# Linearizability
+
+## Summary
+
+Guarantee linearizability in the SQL layer by ensuring that any observed transaction will also be observed by any future transaction.
+Implement this by using a lower-bound timestamp for each timeline.
+Any function that generates a timestamp will either update the lower bound (if it generated a higher timestamp) or use the lower bound.
+
+## Goals
+
+We want to provide linearizability as defined by [Jepsen] which says that
+
+> every operation appears to take place atomically, in some order, consistent with the real-time ordering of those operations
+
+This is limited to transactions (or single statements) in the SQL layer.
+Any statement can use `AS OF` to ignore linearizability, which is only provided when materialize chooses a timestamp.
```

Comment:
An `AS OF` query should probably induce a lower bound on the associated timestamp, right?

The coordinator might want to be involved in the return of data to clients, so that it can e.g. prevent the return of "future" data until the appropriate time arrives. That seems to be an important part of regulating the release of timestamped data back to the client (right now: dataflow does it as soon as it has an answer).

Reply:
> An AS OF query should probably induce a lower bound on the associated timestamp

I would like this behavior. It would bump up the timestamp if the since has advanced past it, yes? I think it's needed because users have no way of controlling compaction well enough to ensure that a future AS OF will work. However it's backward incompatible, so we either need to provide a way for users to tell mz to error (maybe `AS OF EXACT`) or something. I'm not sure if this feature needs to relate to linearizability? Or are you thinking about this in a different way?

> The coordinator might want to be involved in the return of data to clients

Can you describe an example where this is needed? I can't come up with one.

Reply:
As soon as the coordinator sends an event to the `dataflow` layer it probably needs to treat the event as "complete and durable", if it can no longer prevent the information from arriving at the client. One example is that this means that it should not do this before it is certain that all of the inputs to the query are durable. If instead it allowed the results to flow back through it, it could then delay the response until the data are durable, allowing that to be concurrent with the computation.

Alternately, if you wanted that AS OF behavior above, the coordinator would plausibly want to delay the return of the data to avoid ticking forward the time until it needed to.

Generally, an event isn't complete until the response is provided. To the extent that the coordinator is in charge of putting orders on things, it might want to restrict the "event completion" as that is what imposes constraints on e.g. linearizability and strict serializability.